app/src/main/kotlin/compiler/utils/Ref.kt

```diff
@@ -0,0 +1,56 @@
+package compiler.utils
+
+class Ref<out T>(val value: T) {
+    override fun equals(other: Any?): Boolean = other is Ref<*> && other.value === value
+
+    override fun hashCode(): Int = System.identityHashCode(value)
+
+    override fun toString(): String = value.toString()
+}
+
+typealias RefSet<E> = Set<Ref<E>>
+
+typealias RefMap<K, V> = Map<Ref<K>, Ref<V>>
+
+typealias KeyRefMap<K, V> = Map<Ref<K>, V>
+
+typealias MutableRefSet<E> = MutableSet<Ref<E>>
+
+typealias MutableRefMap<K, V> = MutableMap<Ref<K>, Ref<V>>
+
+typealias MutableKeyRefMap<K, V> = MutableMap<Ref<K>, V>
+
```

Comment:
Do we need these at all? I think the explicit types are less confusing

Reply:
Good idea, actually. Removed.